output/cloud/expv2/output.go

```diff
@@ -52,13 +94,75 @@ func (o *Output) Start() error {
 func (o *Output) StopWithTestError(testErr error) error {
 	o.logger.Debug("Stopping...")
 
-	// TODO: add stop implementation
+	o.periodicCollector.Stop()
+	o.periodicFlusher.Stop()
 
 	o.logger.Debug("Stopped!")
 	return nil
 }
 
 // AddMetricSamples receives the samples streaming.
 func (o *Output) AddMetricSamples(s []metrics.SampleContainer) {
-	// TODO: implement
+	select {
+	case <-o.stopSendingMetrics:
+		return
+	default:
+	}
+
+	o.SampleBuffer.AddMetricSamples(s)
+}
+
+// flushMetrics receives a set of metric samples.
+func (o *Output) flushMetrics() {
+	start := time.Now()
+
+	ctx, cancel := context.WithTimeout(context.Background(), o.config.MetricPushInterval.TimeDuration())
+	defer cancel()
+
+	err := o.metricsFlusher.Flush(ctx)
+	if err != nil {
+		o.logger.WithError(err).Error("Failed to push metrics to the cloud")
+
+		if o.shouldStopSendingMetrics(err) {
+			o.logger.WithError(err).Warn("Interrupt sending metrics to cloud due to an error")
+			serr := errext.WithAbortReasonIfNone(
+				errext.WithExitCodeIfNone(err, exitcodes.ExternalAbort),
+				errext.AbortedByOutput,
+			)
+			if o.config.StopOnError.Bool {
+				o.testStopFunc(serr)
+			}
+			close(o.stopSendingMetrics)
+
+			// TODO: just stop the entire output?
+			// otherwise the collector continue the
+			// buffering and it could generate an OOM
+			//
+			// o.Stop()
+		}
+		return
+	}
+
+	o.logger.WithField("t", time.Since(start)).Debug("Successfully flushed buffered samples to the cloud")
+}
+
+// shouldStopSendingMetrics returns true if the output should interrupt the metric flush.
+//
+// note: The actual test execution should continues,
+// since for local k6 run tests the end-of-test summary (or any other outputs) will still work,
+// but the cloud output doesn't send any more metrics.
+// Instead, if cloudapi.Config.StopOnError is enabled
+// the cloud output should stop the whole test run too.
+// This logic should be handled by the caller.
+func (o *Output) shouldStopSendingMetrics(err error) bool {
+	if err == nil {
+		return false
+	}
+	if errResp, ok := err.(cloudapi.ErrorResponse); ok && errResp.Response != nil { //nolint:errorlint
+		// The Cloud service returns the error code 4 when it doesn't accept any more metrics.
+		// So, when k6 sees that, the cloud output just stops prematurely.
+		return errResp.Response.StatusCode == http.StatusForbidden && errResp.Code == 4
+	}
+
+	return false
 }
```

Comment:
Maybe move that to cloudapi package?

Reply:
Considering we want to remove the v1 cloud output implementation in the near future, this will only be used in v2, so no need to move it to a separate library IMO

Reply:
This is now a dedicated PR https://github.com/grafana/k6/pull/3082 let's move the discussion there if you still want it.